### PR TITLE
[BugFix] eliminate constant in group by key

### PR DESCRIPTION
**Pull request by packy92**
## Why I'm doing:
Sql like
`select count(v1), count(distinct v4), 'a' as a from (select * from t1 left join t2 on t1.v1 = t2.v4) t group by a`
may genrate a projection between middle phase agg. The relation is a left outer join relation which changes the nullable info in planfragment builder. This lead a plan like
img_v3_028p_85b5292c-564e-4cf6-840e-837fa0508c4g
This yields a nullable column for output col of agg operation which may cause unexpect behavior.

## What I'm doing:

Fixes #issue
Remove columnRef actually is a constant value in group by keys, if all constant, just keep one constant group by key. The SplitMultiPhaseAggRule#ThreeStageMoreEfficient will make sure we use four stage agg  to process the group by constant key.

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        Map<ColumnRefOperator, ScalarOperator> topMap = Maps.newHashMap();
+        for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : projectOp.getColumnRefMap().entrySet()) {
+            if (aggOp.getGroupingKeys().contains(entry.getKey())) {
+                topMap.put(entry.getKey(), entry.getValue());
+            } else {
+                bottomMap.put(entry.getKey(), entry.getValue());
+            }
+        }
+
+        aggOp.getAggregations().keySet().forEach(e -> topMap.put(e, e));
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java:73]: should check groupby is empty, must save one at least. BTW, I think we need pull all constant value rule? like deriver constant value

Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java:73]: This rule only process all group by keys are constant.  So it's correct to take this scene as a non group by agg.

Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java:73]: Yes, I agree with you. We need a such rule. Let me finish it later.

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java:73]: > This rule only process all group by keys are constant. So it's correct to take this scene as a non group by agg.

yeah, group by a column is differenet with no group by when the rows empty, so need save one at least

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    public boolean check(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator aggOp = input.getOp().cast();
+        LogicalProjectOperator projectOp = input.inputAt(0).getOp().cast();
+        Map<ColumnRefOperator, ScalarOperator> columnRefMap = projectOp.getColumnRefMap();
+        if (aggOp.getGroupingKeys().isEmpty() || aggOp.getAggregations().isEmpty()) {
+            return false;
+        }
+
+        for (ColumnRefOperator groupByCol : aggOp.getGroupingKeys()) {
+            if (!columnRefMap.get(groupByCol).isConstant()) {
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java:52]: What about partial const group by keys? suck as `group by a, b, 1.0, 'a'`? We can always do this optimization once there are constant operators?

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java:52]: How can we distinguish between constant operators and placeholders?
eg:
```
// where 1 reprensents `a`
select a, sum(b) from tbl group by 1;
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java:73]: btw, sr will remove constants value in transformer phase now, you can let it remove all constant in group by, not only `all group by keys are constant`.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java:52]: > How can we distinguish between constant operators and placeholders? eg:
> 
> ```
> // where 1 reprensents `a`
> select a, sum(b) from tbl group by 1;
> ```

will handle it in analyze phase

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java:73]: a other question, why sr will generate `agg(update serialize)-project-agg(update serialize)`? it's a invalid plan...and we should fix the count's nullable info?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java (hunk 2 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateGroupByConstantRule.java]: > What about partial const group by keys? suck as group by a, b, 1.0, 'a'? We can always do this optimization once there are constant operators?
We will handle the constant expr in analyze phase.
This pr mainly focus on group by a column but this column actually is a constant.

**Review by LiShuMing — DISMISSED** [review]

**Review by Seaven — APPROVED** [review]

**Review by LiShuMing — APPROVED** [review]

**Pull request merged**